### PR TITLE
Add attribute to handle deserialization of constructor parameters with name overrides

### PR DESCRIPTION
**Pull request by praneetloke**
This PR is in support of https://github.com/pulumi/pulumi/issues/14130.

I have a local override for the native provider in question to use the changes in this PR, as well as codegen changes in pulumi/pulumi using the PR https://github.com/pulumi/pulumi/pull/14246 with additional changes to emit the attribute introduced in this PR.

Here's how it looks in practice in a provider's SDK:

```csharp
    [OutputType]
    public sealed class ListOneClicksProperties
    {
        public readonly ImmutableArray<Outputs.OneClicks> _1Clicks;

        [OutputConstructor]
        private ListOneClicksProperties([OutputConstructorParameterAttribute("1_clicks")] ImmutableArray<Outputs.OneClicks> _1Clicks)
        {
            this._1Clicks = _1Clicks;
        }
    }
```

### CODE REVIEW

File: sdk/Pulumi/Serialization/Attributes.cs
Comment (justinvp) [inline sdk/Pulumi/Serialization/Attributes.cs:117]: The type should be `sealed`.

Also, I wonder if a better name would be `OutputConstructorParameterAttribute`, to align with `OutputConstructorAttribute`:

```suggestion
    public sealed class OutputConstructorParameterAttribute : Attribute
```

I also wonder if this should be moved directly below `OutputConstructorAttribute` (line 76).

File: sdk/Pulumi/Serialization/Attributes.cs
Comment (justinvp) [inline sdk/Pulumi/Serialization/Attributes.cs:124]:
```suggestion
        public ConstructorParameterNameOverrideAttribute(string name)
        {
            Name = name;
        }
```

File: sdk/Pulumi/Serialization/Converter.cs
Comment (justinvp) [inline sdk/Pulumi/Serialization/Converter.cs:206]: Nit

```suggestion
                if (attribute != null)
```

**Review by justinvp — COMMENTED** [review]
Thanks @praneetloke! Some comments.

Also, this needs tests. See https://github.com/pulumi/pulumi-dotnet/blob/8f1dd5e358a18aa8ac3f034dcb22818f2fa77e36/sdk/Pulumi.Tests/Serialization/ComplexTypeConverterTests.cs#L82

Your usage in the PR description doesn't quite look how I'd expect it to. I'd expect something like:

```c#
    [OutputType]
    public sealed class ListOneClicksProperties
    {
        public readonly ImmutableArray<Outputs.OneClicks> _1Clicks;

        [OutputConstructor]
        private ListOneClicksProperties([OutputConstructorParameter("1_clicks")] ImmutableArray<Outputs.OneClicks> _1Clicks)
        {
            this._1Clicks = _1Clicks;
        }
    }
```

File: sdk/Pulumi/Serialization/Attributes.cs
Comment (praneetloke) [reply sdk/Pulumi/Serialization/Attributes.cs]: I've made changes for both suggestions you've made here. Also on the topic of the attribute class being `sealed` I noticed that `ResourceTypeAttribute` is not marked as such. That's the one I was using as reference and hence didn't mark the new one as `sealed` initially. Would you like me to mark it as `sealed` as well?

**Comment by praneetloke** [comment]
@justinvp I've pushed new commits with changes based on your feedback.

> Your usage in the PR description doesn't quite look how I'd expect it to. I'd expect something like:

Ah, indeed. I made a mistake in the codegen and wrote out the attribute on the class member instead of the constructor param...despite the name of the attribute 🤦‍♂️ . I'll fix that when I open the PR in `pulumi/pulumi` though. I've noted it down.

**Comment by praneetloke** [comment]
Bump.

**Review by justinvp — APPROVED** [review]
LGTM! Thank you @praneetloke and apologies for the long delay getting this reviewed!

**Comment by praneetloke** [comment]
@justinvp awesome! The status checks for this PR still says that the workflow requires approval from a maintainer though and haven't run. Any idea why that might be?

**Comment by Frassle** [comment]
Looks like you just need to run "dotnet run format-sdk"

**Comment by praneetloke** [comment]
Ah got it. I'll remember that for next time. Thanks, @Frassle.

**Pull request merged**
